README.md
# jfx_wrapper

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-04-14
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-04-14
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/jfx_wrapper/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/jfx_wrapper/compare/0.1.0...0.1.1


test/de/trema02/jfx_wrapper/core_test.clj
(ns de.trema02.jfx-wrapper.core-test
  (:require [clojure.test :refer :all]
            [de.trema02.jfx-wrapper.core :as c]))

(deftest explain-app-data-error-test
  (testing "explain-app-data-error"
    (let [gui (de.trema02.jfx_wrapper.core.GUI.)
          data (c/get-app-data)]
      (is (= "Not a map." (c/explain-app-data-error gui data :map)))
      (is (= "Undefined keys: #{:title}." (c/explain-app-data-error gui (dissoc data :title) :all-keys)))
      (is (= ":config-fn must be nil or a function." (c/explain-app-data-error gui data :config-fn)))
      (is (= ":title must be a string." (c/explain-app-data-error gui data :title)))
      (is (= ":mk-root-fn must be nil or a function." (c/explain-app-data-error gui data :mk-root-fn)))
      (is (= ":css-name must be nil or a valid resource." (c/explain-app-data-error gui data :css-name)))
      (is (= ":x-size must be a positive integer." (c/explain-app-data-error gui data :x-size)))
      (is (= ":y-size must be a positive integer." (c/explain-app-data-error gui data :y-size)))
      (is (= ":image-name must be nil or a valid resource." (c/explain-app-data-error gui data :image-name)))
      (is (= ":fxml-name must be nil or a valid resource." (c/explain-app-data-error gui data :fxml-name)))
      (is (= ":end-fn must be nil or a function." (c/explain-app-data-error gui data :end-fn)))
      (is (= "OK." (c/explain-app-data-error gui data :ok))))))

(deftest get-app-data-test
  (testing "get-app-data"
    (is (= {:end-fn nil, :config-fn nil, :title "", :mk-root-fn nil, :css-name nil, :y-size 480, :image-name nil, :x-size 640, :fxml-name nil}
           (c/get-app-data)))))

(deftest app-data-valid?-test
  (testing "app-data-valid?"
    (let [gui (de.trema02.jfx_wrapper.core.GUI.)
          data (c/get-app-data)]
      (is (= :map (c/app-data-valid? gui 0)))
      (is (= :all-keys (c/app-data-valid? gui (dissoc data :title))))
      (is (= :config-fn (c/app-data-valid? gui (assoc data :config-fn 0))))
      (is (= :title (c/app-data-valid? gui (assoc data :title 0))))
      (is (= :mk-root-fn (c/app-data-valid? gui (assoc data :mk-root-fn 0))))
      (is (= :css-name (c/app-data-valid? gui (assoc data :css-name 0))))
      (is (= :x-size (c/app-data-valid? gui (assoc data :x-size -1))))
      (is (= :y-size (c/app-data-valid? gui (assoc data :y-size -1))))
      (is (= :image-name (c/app-data-valid? gui (assoc data :image-name 0))))
      (is (= :fxml-name (c/app-data-valid? gui (assoc data :fxml-name 0))))
      (is (= :end-fn (c/app-data-valid? gui (assoc data :end-fn 0))))
      (is (= :ok (c/app-data-valid? gui data))))))

;(run-tests)


src/de/trema02/jfx_wrapper/core.clj
(ns de.trema02.jfx-wrapper.core
  (:gen-class)
  (:import [javafx.application Application Platform]
           [javafx.stage Stage]
           [javafx.fxml FXMLLoader]
           [javafx.event EventHandler]
           [javafx.scene Scene]
           [javafx.scene.image Image])
  (:require [clojure.set :as st]))

(gen-class
   :name de.trema02.jfx_wrapper.core.GUI
   :extends javafx.application.Application
   :main false
   :exposes-methods {stop stop_gui}
   :methods [[entry [] void]]) ;"abstract methods work automagically"


(def mc (atom nil))
(def gui-root (atom nil))
(def gui (atom nil))
(def application (atom nil))
(def style-sheet (atom nil))
(def image (atom nil))
(def ^{:private false} primary-stage (atom nil))
(def data (atom {:x-size 640 :y-size 480 :fxml-name nil :css-name nil
                 :image-name nil :config-fn nil :title "" :mk-root-fn nil
                 :end-fn nil :debug false}))
(def all-keys '(:config-fn :title :mk-root-fn :css-name :x-size :y-size
                :image-name :fxml-name :end-fn :debug))

(defn app-data-valid? [app mp]
  (let [result (atom :map)]
    (and
     (map? mp)
     (do (reset! result :all-keys) true)
     (every? #(contains? mp %)
             all-keys)
     (do (reset! result :config-fn) true)
     (let [val (:config-fn mp)]
       (or (nil? val) (fn? val)))
     (do (reset! result :title) true)
     (let [val (:title mp)]
       (string? val))
     (do (reset! result :mk-root-fn) true)
     (let [val (:mk-root-fn mp)]
       (or (nil? val) (fn? val)))
     (do (reset! result :css-name) true)
     (let [val (:css-name mp)]
       (or (nil? val) (and (string? val) (.. app getClass (getResource val)))))
     (do (reset! result :x-size) true)
     (let [val (:x-size mp)]
       (pos? val))
     (do (reset! result :y-size) true)
     (let [val (:y-size mp)]
       (pos? val))
     (do (reset! result :image-name) true)
     (let [val (:image-name mp)]
       (or (nil? val) (and (string? val) (.. app getClass (getResource val)))))
     (do (reset! result :fxml-name) true)
     (let [val (:fxml-name mp)]
       (or (nil? val) (and (string? val) (.. app getClass (getResource val)))))
     (do (reset! result :end-fn) true)
     #_(let [val (:end-fn mp)] ; cannot test handler
       (or (nil? val) (fn? val)))
     (do (reset! result :ok) true))
    @result))

(defn explain-app-data-error [_ mp key]
  (case key
    :map (str "Not a map.")
    :all-keys (let [all-key-set (into #{} all-keys)
                    mp-key-set (into #{} (keys mp))]
                (str "Undefined keys: "
                     (st/difference all-key-set mp-key-set)
                     "."))
    :config-fn (str ":config-fn must be nil or a function.")
    :title (str ":title must be a string.")
    :mk-root-fn (str ":mk-root-fn must be nil or a function.")
    :css-name (str ":css-name must be nil or a valid resource.")
    :x-size (str ":x-size must be a positive integer.")
    :y-size (str ":y-size must be a positive integer.")
    :image-name (str ":image-name must be nil or a valid resource.")
    :fxml-name (str ":fxml-name must be nil or a valid resource.")
    :end-fn (str ":end-fn must be a handler.")
    :ok (str "OK.")))

(defn set-app-data [app mp]
  (let [result-key (app-data-valid? app mp)
        ok (= result-key :ok)]
    (if ok
      (do
        (reset! application app)
        (reset! data mp))
      (throw (Exception. (str "jfx_wrapper.core/set-app-data: " (explain-app-data-error app mp result-key)))))))

(defn get-app-data []
  @data)

(defn event-handler [f & args]
  (proxy [EventHandler] []
    (handle [event] (apply f event args))))

(declare start-stage)
(def close-handler (event-handler (fn [_]
                                    (if (:debug @data)
                                      (Platform/runLater #(start-stage))
                                      (Platform/exit)))))

(defn start-stage []
  (let [{:keys [fxml-name mk-root-fn]} @data]
    (cond
      fxml-name (let [res (.. @gui getClass (getResource fxml-name))
                      loader (FXMLLoader. res)
                      r (.load loader)
                      main-controller (.getController loader)]
                  (reset! gui-root r)
                  (reset! mc main-controller))
      mk-root-fn (let [r (mk-root-fn)]
                   (reset! gui-root r))
      :else (throw (Exception. (str "start: :fxml-name or :mk-root-fn must be specified."))))
    (let [css (:css-name @data)]
      (if css
        (reset! style-sheet (.. @gui getClass (getResource css) toExternalForm))
        nil))
    (let [scene (Scene. @gui-root (:x-size @data) (:y-size @data))
          stage (Stage.)]
      (let [img (:image-name @data)]
        (if img
          (do
            (reset! image (Image. (.. @application getClass (getResourceAsStream img))))
            (doto (.getIcons stage) (.add @image)))
          nil))
      (if @style-sheet
        (.. scene getStylesheets (add @style-sheet))
        nil)
      (.setTitle stage (:title @data))
      (doto stage
        (.setScene scene)
        (.setOnCloseRequest (or (:end-fn @data) close-handler)))
      (let [cf (:config-fn @data)]
        (if cf
          (cf stage @mc)
          nil))
      (.show stage))))

(defn setup [data]
  (let [gui (de.trema02.jfx_wrapper.core.GUI.)]
    (set-app-data gui data)
    (.entry gui)))

(defn -start [this stage]
  (reset! primary-stage stage)
  (reset! gui this)
  (Platform/setImplicitExit false)
  (Platform/runLater #(start-stage))
  nil)

(defn- -entry [this]
  (Application/launch (.getClass this) (make-array java.lang.String 0)))

(defn -stop [this]
  (println "STOP")
  (.stop_gui this))

(comment defn -main [& _]
  (.entry (de.trema02.jfx_wrapper.core.GUI.)))
